src/common/signalk/SKNMEAVisitor.cpp

```diff
@@ -71,9 +73,90 @@ void SKNMEAVisitor::visitSKNavigationAttitude(const SKUpdate &u, const SKPath &p
 };
 
 void SKNMEAVisitor::visitSKNavigationHeadingMagnetic(const SKUpdate &u, const SKPath &p, const SKValue &v) {
-  NMEASentenceBuilder sb2("II", "HDM", 2);
-  sb2.setField(1, SKRadToDeg(v.getNumberValue()), 1);
-  sb2.setField(2, "M");
+  NMEASentenceBuilder sb("II", "HDM", 2);
+  sb.setField(1, SKRadToDeg(v.getNumberValue()), 1);
+  sb.setField(2, "M");
 
-  _sentences.add(sb2.toNMEA() + "\r\n");
+  _sentences.add(sb.toNMEA() + "\r\n");
 };
+
+// ***********************  Wind Speed and Angle  ************************
+//  Talker ID: WI Weather Instruments
+//             VW Velocity Sensor, Mechanical?
+//  also seen: IIVWR
+//
+//  NMEA0183  MWV Wind Speed and Angle  --> recommanded
+//              1  2  3  4
+//              |  |  |  |
+//      $--MWV,x.x,a,x.x,a*hh
+//  1) Wind Angle, 0 to 360 degrees
+//  2) Reference, R = Relative, T = True
+//  3) Wind Speed
+//  4) Wind Speed Units: K = km/s, M = m/s, N = Knots
+//  5) Status, A = Data Valid
+//
+//      VWR Relative Wind Speed and Angle  --> deprecated
+//              1  2  3  4  5  6  7  8
+//              |  |  |  |  |  |  |  |
+//      $--VWR,x.x,a,x.x,N,x.x,M,x.x,K*hh
+//  1) Wind direction magnitude in degrees
+//  2) Wind direction Left/Right of bow
+//  3) Speed
+//  4) N = Knots
+//  5) Speed
+//  6) M = Meters Per Second
+//  7) Speed
+//  8) K = Kilometers Per Hour
+// ************************************************************************
+void SKNMEAVisitor::visitSKEnvironmentWindAngleApparent(const SKUpdate &u, const SKPath &p, const SKValue &v) {
+
+  float windAngle;
+  float windSpeed;
+  windAngle = v.getNumberValue();
+
+  // Validation
+  if ((windAngle < 0)||(windAngle > 2 * M_PI)||(!u.hasEnvironmentWindSpeedApparent())) return;
```

Comment:
No. The values we generate when parsing NMEA2000 data will have an angle between -PI and +PI if it's an angle and not a direction. But here we would refuse this as an input and not generate an NMEA sentence for it.

Reply:
Why are you validating that we `hasEnvironmentWindSpeedApparent()` here but not on the other ones? I think we should do the same thing everywhere. Here it's a private method called by another method who does the `if()` for us so I think we can skip it.

Reply:
Good question!
Should we do some validations, or not?
of course they will have to work, not like mine.... ;-)

I was not clear here, this is why I sometimes do and sometimes not, more or less to discuss it.
May be just to check !N2kIsNA() or even nothing at all?
What is your opinion?

All the rest I will read carefully and do it later...

Reply:
For the ´!u.hasEnvironmentWindSpeedApparent()´ validation:

**My logic was:**
For this sentence we need two different Signal K.
one is triggered by the visitor/update itself, here the angle.
The other one needed, I check.

**But may be it should the other way round:**
1. incoming PGN contents speed & angle
2. it get's parsed and it is written 
´´´
             update->setEnvironmentWindSpeedOverGround(windSpeed);
             update->setEnvironmentWindDirectionTrue(windAngle);
´´´
3. The NMEA0183 writing function has to wait for update on Angle and then it can take the new speed value for this sentence. 

If this logic is right, then we have to carefully watch, which value is the last coming in and this one triggers the following calculations.

Reply:
You are absolutely right! Sorry! Yes you do need the `!u.hasEnvironmentWindSpeedApparent()` validation.
My logic is that if the data comes through one PGN or one NMEA sentence, all the information should be transformed into one SignalK Delta message (one SKUpdate object). This is what the official nmea2000 tool and iKommunicate do.

For the angles, I think we should try to be as tolerant as possible and accept values that are not normalized. We should always send normalized values (0 to 2pi for directions, -pi to pi for angles).

Reply:
Good morning :-)

May be a function which normalizes angles and calculate +/- angle, something like:
```
void directionToAngle(double &angle) {
    NORMALIZE_ANGLE(angle);
    if (angle > M_PI ) angle += - M_2xPI;
}
```

Reply:
good morning :) 

As I said in my other big comment, I think you were right and we should be strict. Better detect problems early than later. Sorry for changing my mind :disappointed: 

Thanks for a great discussion that led me to change my mind! :1st_place_medal: